File: ros2_ws/src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"


class AddTwoIntsClientNode : public rclcpp::Node 
{


public:
    AddTwoIntsClientNode(): Node("add_two_ints_client") 
    {
        
        // thread1_ = std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService,this,1,4 ));

        threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService,this,1,4 )));
        threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService,this,9,4 )));



    }


    void callAddTwoIntsService(int a , int b)
    {
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), " WAiting for server to be up");
        }
        
        auto request = std:: make_shared<example_interfaces::srv::AddTwoInts::Request>();
        
        request->a = a;
        request->b = b;
        
        auto future = client->async_send_request(request);

        
        try
        {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), " %d + %d = %d " , a , b, (int)response->sum);
        }
        catch(const std::exception& e)
        {
            RCLCPP_ERROR(this->get_logger() , "Service call failed");
        }
        
        // future.get() will wait until we have a response for the future
        // It can also throw exception if there is any error

    }

private:

    // std::thread thread1_;
    std::vector<std::thread> threads_;
    

};

int main(int argc , char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}







// we start the node, it creates client, starts service, creates request, 
// we wait for response, future will get. This will br=lock the thread here and that means that the function will not exist

File: ros2_ws/src/my_py_pkg/my_py_pkg/add_two_ints_client.py
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial

class AddTwoIntsClientNode(Node): # MODIFY NAME

    def __init__(self):
        super().__init__("add_two_ints_client") # MODIFY NAME
        self.call_add_two_ints_server(6,7)


    def call_add_two_ints_server(self,a,b):
        # we will create a cliet, we create the request, 
        # we call the server and also then see how we can wait fot the response 
        

        # we are in the class so self.
        client = self.create_client(AddTwoInts , "add_two_ints")

    # wait for the server

        while not client.wait_for_service(1.0):
            # it waits for 1.0 sec. if time is not given it waits forever

            self.get_logger().warn("Waiting for server Add For Two Ints ....")
        request = AddTwoInts.Request()

        request.a = a
        request.b = b

        future = client.call_async(request)

        # wait for future
        future.add_done_callback(partial(self.callback_call_add_two_ints , a=a , b=b))


    def callback_call_add_two_ints(self ,future , a,b):

        try:
            response = future.result()
            self.get_logger().info(str(a) + "+" + str(b) + "=" + str(response.sum))
        except Exception as e:
            self.get_logger().error("Service call failed %r" %(e,))








def main(args = None):

    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

    